lib/buffer.c

```diff
@@ -38,17 +40,19 @@ void buffer_free(void *struct_buffer)
 
 struct buffer *buffer_copy(const void *data, size_t data_len)
 {
-	struct buffer *buf;
-	buf = malloc(sizeof(*buf));
+	struct buffer *buf = calloc(1, sizeof(struct buffer));
+
 	if (!buf)
 		goto err_out;
 
-	buf->p = malloc(data_len);
-	if (!buf->p)
-		goto err_out_free;
+	if (data_len > 0) {
+		buf->p = malloc(data_len);
+		if (!buf->p)
+			goto err_out_free;
 
-	memcpy(buf->p, data, data_len);
-	buf->len = data_len;
+		memcpy(buf->p, data, data_len);
+		buf->len = data_len;
+	} else buf->p = NULL;
```

Comment:
Code is correct; however, check callers in-tree for the possibility of being coded with the assumption that malloc(0) results in a valid pointer.  This code includes the behavior change that a zero-length buffer now results in NULL.  Must check that callers support this behavior change.